src/planscape/forsys/forsys_request_params.py

```diff
@@ -0,0 +1,294 @@
+from plan.views import (get_plan_by_id, get_user)
+
+
+# A list of coordinates representing a polygon.
+class PolygonCoordinatesFromUrlParams(TypedDict):
+    coordinates: list[tuple[float, float]]
+
+
+# A shape composed of multiple disjoint polygons.
+# Accompanied by SRID and an ID denoting project or plan ID.
+class GeoFromUrlParams(TypedDict):
+    # Project or Plan ID.
+    id: int | None
+    # SRID
+    srid: int
+    # Disjoint polygons that are part of the project area.
+    polygons: list[PolygonCoordinatesFromUrlParams]
+
+
+def _read_common_url_params(self, params: QueryDict) -> None:
+    self.region = params.get(
+        self._URL_REGION, self._DEFAULT_REGION)
+    self.priorities = params.getlist(
+        self._URL_PRIORITIES, self._DEFAULT_PRIORITIES)
+    self.priority_weights = params.getlist(
+        self._URL_PRIORITY_WEIGHTS,
+        [1 for i in range(len(self.priorities))])
+    self.priority_weights = [float(pw) for pw in self.priority_weights]
+    if len(self.priorities) != len(self.priority_weights):
+        raise AssertionError(
+            "expected %d priority weights, instead, %d were given" %
+            (len(self.priorities),
+                len(self.priority_weights)))
+
+
+def _check_geo_from_url_params_fields_exist(
+        geo: GeoFromUrlParams, field_name: str) -> None:
+    if "polygons" not in geo.keys():
+        raise Exception(
+            'url parameter, %s, missing field, "polygons"' % field_name)
+    if len(geo["polygons"]) == 0:
+        raise Exception(
+            'url parameter, %s, field, "polygons" is an empty list' %
+            (field_name))
+    if 'id' not in geo.keys():
+        raise Exception('url params, %s, missing field, "id"' % field_name)
+
+
+def _transform_geo_from_url_params_into_multipolygon(
+        geo: GeoFromUrlParams, field_name: str) -> MultiPolygon:
+    _check_geo_from_url_params_fields_exist(geo, field_name)
+    srid = 4269 if 'srid' not in geo.keys(
+    ) else geo['srid']  # TODO: make 4269 a constant.
+    polygons: list[Polygon] = []
+    for p in geo['polygons']:
+        polygon = Polygon(tuple(p['coordinates']))
+        polygon.srid = srid
+        if not polygon.valid:
+            raise Exception(
+                "polygon described by %s is invalid - %s" %
+                (json.dumps(geo), polygon.valid_reason))
+        polygons.append(polygon)
+    if len(polygons) == 0:
+        raise Exception(
+            "multipolygon described by %s missing polygons" % json.dumps(
+                project_area))
+    m = MultiPolygon(polygons)
+    m.srid = srid
+    return m
+
+
+# Gathers forsys input parameters from url params, database lookups, or a
+# combination of the two.
+# Of note, the option to set all forsys input paramters via url parameters is
+# intended for backend debugging purposes while the option to set most forsys
+# input parameters via database lookups is intended for production.
+class ForsysProjectAreaRankingRequestParams():
```

Comment:
i wonder if we could make this shorter. maybe ForsysRankingParams here and ForsysGenerationParams below? it might make referencing this object more readable

Reply:
Done.
Renamed to ForsysRankingRequestParams and ForsysGenerationRequestParams.
Keep the "Request" term since it roughly associates with HttpRequest, and I want to differentiate it from ForsysRankingInput.